### PR TITLE
1846 support solargraph lsp

### LINKED ISSUE #1846: Add solargraph language server for ruby.
Would be nice if ALE supported  [Solargraph](https://solargraph.org/) that is a language server for ruby.

### PR DESCRIPTION
**Pull request by hsanson**
Adds support for solargraph language server.

Closes #1846

### CODE REVIEW

File: ale_linters/ruby/solargraph.vim [hunk 1]
```diff
@@ -0,0 +1,19 @@
+" Author: Horacio Sanson - https://github.com/hsanson
+" Description: Solargraph Language Server https://solargraph.org/
+
+call ale#Set('ruby_solargraph_host', '127.0.0.1')
+call ale#Set('ruby_solargraph_port', '7658')
+
+function! ale_linters#ruby#solargraph#GetAddress(buffer) abort
+    let l:host = ale#VarFunc('ruby_solargraph_host')(a:buffer)
+    let l:port = ale#VarFunc('ruby_solargraph_port')(a:buffer)
```
Comment (w0rp) [inline ale_linters/ruby/solargraph.vim]: Use `ale#Var(a:buffer, variable_name)` instead. `VarFunc` is just for writing simple callbacks easily.

File: autoload/ale/ruby.vim [hunk 2]
```diff
@@ -20,3 +20,65 @@ function! ale#ruby#FindRailsRoot(buffer) abort
...
+    \)
+
+    if isdirectory(l:dir)
+      return l:dir
+    endif
+
+    let l:dir = fnamemodify(
+    \   ale#path#FindNearestFile(a:buffer, "Rakefile"),
+    \   ':h'
+    \)
```
Comment (w0rp) [inline autoload/ale/ruby.vim]: You're searching for the same two files twice here.

File: autoload/ale/ruby.vim [hunk 3]
```diff
@@ -20,3 +20,65 @@ function! ale#ruby#FindRailsRoot(buffer) abort
...
+
+    if isdirectory(l:dir)
+      return l:dir
+    endif
```
Comment (w0rp) [inline autoload/ale/ruby.vim]: Add `return ''` to the end, or the function implicitly returns `0`.

**Comment by hsanson** [comment]
@w0rp I have added tests, simplified the find root logic, and updated the documentation as requested.

BTW: the dev documentation does not mention anything about rebase policies. I rebased master and squashed everything into a single commit for easier review. Hope this is not a problem as I had to force push for this to work.

**Comment by w0rp** [comment]
There is no policy. I just merge code that's good and use `--squash` if the commits are messy.

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by w0rp** [comment]
I might add an option to disable TCP linters by default later if it turns out that trying and failing to make connections causes an issue. We'll see what happens.

**Pull request merged**